src/function/scalar/system/aggregate_export.cpp

```diff
@@ -0,0 +1,304 @@
+			result_ptr[i] =
+			    StringVector::AddStringOrBlob(result, (const char *)state1.GetDataUnsafe(), bind_data.state_size);
+			continue;
+		}
+
+		// we actually have to combine
+		D_ASSERT(state0.GetSize() == bind_data.state_size);
+		D_ASSERT(state1.GetSize() == bind_data.state_size);
+
+		memcpy(local_state.state_buffer0.get(), state0.GetDataUnsafe(), bind_data.state_size);
+		memcpy(local_state.state_buffer1.get(), state1.GetDataUnsafe(), bind_data.state_size);
+
+		bind_data.aggr.combine(local_state.state_vector0, local_state.state_vector1, 1);
+
+		result_ptr[i] =
+		    StringVector::AddStringOrBlob(result, (const char *)local_state.state_buffer1.get(), bind_data.state_size);
+	}
+}
+
+static unique_ptr<FunctionData> BindAggregateState(ClientContext &context, ScalarFunction &bound_function,
+                                                   vector<unique_ptr<Expression>> &arguments) {
+
+	// grab the aggregate type and bind the aggregate again
+
+	// the aggregate name and types are in the logical type of the aggregate state, make sure its sane
+	auto &arg_return_type = arguments[0]->return_type;
+
+	if (arg_return_type.id() != LogicalTypeId::AGGREGATE_STATE) {
+		throw BinderException("Can only FINALIZE aggregate state, not %s", arg_return_type.ToString());
+	}
+	// combine
+	if (arguments.size() == 2 && arguments[0]->return_type != arguments[1]->return_type) {
+		throw BinderException("Cannot COMBINE aggregate states from different functions, %s <> %s",
+		                      arguments[0]->return_type.ToString(), arguments[1]->return_type.ToString());
+	}
+
+	// following error states are only reachable when someone messes up creating the state_type which is impossible from
+	// SQL
+
+	auto state_type = AggregateStateType::GetStateType(arg_return_type);
+
+	// now we can look up the function in the catalog again and bind it
+	auto func = Catalog::GetCatalog(context).GetEntry(context, CatalogType::SCALAR_FUNCTION_ENTRY, DEFAULT_SCHEMA,
+	                                                  state_type.function_name);
+	if (func->type != CatalogType::AGGREGATE_FUNCTION_ENTRY) {
+		throw InternalException("Could not find aggregate %s", state_type.function_name);
+	}
+	auto aggr = (AggregateFunctionCatalogEntry *)func;
+
+	string error;
+	idx_t best_function = Function::BindFunction(aggr->name, aggr->functions, state_type.bound_argument_types, error);
+	if (best_function == DConstants::INVALID_INDEX) {
+		throw InternalException("Could not re-bind exported aggregate %s: %s", state_type.function_name, error);
+	}
+	auto &bound_aggr = aggr->functions[best_function];
+	if (bound_aggr.return_type != state_type.return_type || bound_aggr.arguments != state_type.bound_argument_types) {
+		throw InternalException("Type mismatch for exported aggregate %s", state_type.function_name);
+	}
+
+	// maybe we have to call the bind callback for the function, too
+	unique_ptr<FunctionData> aggr_bind;
+	if (bound_aggr.bind) {
```

Comment:
Hm, this is actually a slight complication: the bind can be used in certain cases to influence how the function behaves. For example, the quantile function takes the % as a constant expression that will then change what it does in the finalize call.

Perhaps we need to add a `Serialize/Deserialize` to the bind data to allow this bind data state to be stored inside the `AggregateState` type, and avoid having to re-call the bind?

Reply:
Let's discuss tomorrow not sure what the best way to address this is

Reply:
Checking for presence of custom binder now and throwing an error if so to defend against this